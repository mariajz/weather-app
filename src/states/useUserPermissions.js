import { atom, useAtom } from 'jotai';

const locationPermissionAtom = atom('unavailable');

const useUserPermissions = () => {
    const [locationPermission, setLocationPermission] = useAtom(
        locationPermissionAtom,
    );

    return {
        locationPermission,
        setLocationPermission,
    };
};
export default useUserPermissions;
