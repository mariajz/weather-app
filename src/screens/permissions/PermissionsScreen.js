import { useNavigation } from '@react-navigation/native';
import React, { useEffect } from 'react';
import { Text } from 'react-native';
import useGetPermissions from '../../hooks/useGetPermissions';
import useUserPermissions from '../../states/useUserPermissions';
import { PermissionsWrapper } from './PermissionsScreen.style';

const PermissionsScreen = () => {
    const { requestLocationPermission } = useGetPermissions();
    const { locationPermission } = useUserPermissions();
    const navigation = useNavigation();

    useEffect(() => {
        if (locationPermission === 'unavailable') {
            requestLocationPermission();
        }
    }, [locationPermission, requestLocationPermission]);

    useEffect(() => {
        if (locationPermission === 'granted') {
            navigation.navigate({
                name: 'ContentLoaderScreen',
                params: {},
            });
        }
    });

    return (
        <>
            <If condition={locationPermission === 'denied'}>
                <PermissionsWrapper>
                    <Text testID="text">
                        To continue, please grant location permission. You can
                        do this by going to your device's settings and enabling
                        location access for the app.
                    </Text>
                </PermissionsWrapper>
            </If>
        </>
    );
};

export default PermissionsScreen;
