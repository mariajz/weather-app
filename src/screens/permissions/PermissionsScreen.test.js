import { render } from '@testing-library/react-native';
import React from 'react';
import useUserPermissions from '../../states/useUserPermissions';
import PermissionsScreen from './PermissionsScreen';

const mockNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
    useNavigation: () => ({
        navigate: mockNavigate,
    }),
}));

jest.mock('../../states/useUserPermissions');

const mockRequestLocationPermission = jest.fn();
jest.mock('../../hooks/useGetPermissions', () => () => ({
    requestLocationPermission: mockRequestLocationPermission,
}));

describe('PermissionsScreen', () => {
    beforeEach(() => {
        useUserPermissions.mockImplementation(() => ({
            locationPermission: 'unavailable',
        }));
    });
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render PermissionsScreen', () => {
        const container = render(<PermissionsScreen />);

        expect(container).toMatchSnapshot();
    });

    it('should navigate to content loader screen when locationPermission is granted', async () => {
        useUserPermissions.mockImplementation(() => ({
            locationPermission: 'granted',
        }));
        render(<PermissionsScreen />);

        expect(mockNavigate).toHaveBeenCalledWith({
            name: 'ContentLoaderScreen',
            params: {},
        });
    });

    it('should not navigate to content loader screen and show error message when locationPermission is denied', async () => {
        useUserPermissions.mockImplementation(() => ({
            locationPermission: 'denied',
        }));
        const { getByTestId } = render(<PermissionsScreen />);

        expect(mockNavigate).toHaveBeenCalledTimes(0);
        expect(getByTestId('text').props.children).toStrictEqual(
            "To continue, please grant location permission. You can do this by going to your device's settings and enabling location access for the app.",
        );
    });
});
