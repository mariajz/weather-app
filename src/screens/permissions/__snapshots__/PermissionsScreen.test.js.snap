// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsScreen should render PermissionsScreen 1`] = `null`;
